targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_hal_pcd.c

```diff
@@ -1042,26 +1027,84 @@ HAL_StatusTypeDef HAL_PCD_EP_Transmit(PCD_HandleTypeDef *hpcd, uint8_t ep_addr,
   __HAL_UNLOCK(&hpcd->EPLock[ep_addr & 0x7F]);
   return HAL_OK;
 }
+
 /**
   * @brief  Abort a transaction.  
   * @param  hpcd: PCD handle
   * @param  ep_addr: endpoint address
-  * @param  pBuf: pointer to the transmission buffer   
-  * @param  len: amount of data to be sent
+  * @note IN endpoints must have NAK enabled before calling this function
+  * @note OUT endpoints must have global out NAK enabled before calling this
+  *           function. Furthermore, the RX fifo must be empty or the status
+  *           HAL_BUSY will be returned.
   * @retval HAL status
   */
 HAL_StatusTypeDef HAL_PCD_EP_Abort(PCD_HandleTypeDef *hpcd, uint8_t ep_addr)
 {
-  HAL_StatusTypeDef ret;
+  USB_OTG_GlobalTypeDef *USBx = hpcd->Instance;
+  HAL_StatusTypeDef ret = HAL_OK;
+  uint32_t count = 0;
   USB_OTG_EPTypeDef *ep;
 
-  ep = &hpcd->IN_ep[ep_addr & 0x7F];
+  if ((0x80 & ep_addr) == 0x80)
+  {
+    ep = &hpcd->IN_ep[ep_addr & 0x7F];
+  }
+  else
+  {
+    ep = &hpcd->OUT_ep[ep_addr];
+  }
 
-  /*setup and start the Xfer */
   __HAL_LOCK(&hpcd->EPLock[ep_addr & 0x7F]);
-   ret = USB_EPStopXfer(hpcd->Instance , ep, hpcd->Init.dma_enable); 
+
+  ep->num   = ep_addr & 0x7F;
+  ep->is_in = ((ep_addr & 0x80) == 0x80);
+
+  USB_EPSetNak(hpcd->Instance, ep);
+
+  if ((0x80 & ep_addr) == 0x80)
+  {
+      ret = USB_EPStopXfer(hpcd->Instance , ep);
+      if (ret == HAL_OK)
+      {
+        ret = USB_FlushTxFifo(hpcd->Instance, ep_addr & 0x7F);
+      }
+  }
+  else
+  {
+    /* Set global NAK */
+    USBx_DEVICE->DCTL |= USB_OTG_DCTL_SGONAK;
+
+    /* Read all entries from the fifo so global NAK takes effect */
+    while (__HAL_PCD_GET_FLAG(hpcd, USB_OTG_GINTSTS_RXFLVL))
+    {
+      PCD_ReadRxFifo(hpcd);
+    }
+
+    /* stop the transfer */
+    while (count++ < 200000U)
```

Comment:
Can you use a #define for this value ? And also add a comment explaining the approx timeout delay if you know it ? 

Why you didn't use the same timeout method as done in other ST HAL files (i.e. using HAL_GetTick) ?

Reply:
I'll add a comment here. In theory this loop should should execute at most once and that is when a setup packet is received. After the setup packet has been received the data (or status) phase of the control transfer can be will be naked which prevents further packet reception. Maybe its best to update this to remove the loop entirely.

I used a while loop with a count rather than HAL_GetTick to keep consistent with the other USB code (see snippit below). I can update this to use HAL_GetTick if you would like. 
https://github.com/ARMmbed/mbed-os/blob/f3424da060734fa9e27bf6a798747fb48ad63938/targets/TARGET_STM/TARGET_STM32F2/device/stm32f2xx_ll_usb.c#L335-L342

Reply:
No that's ok. Since the RTC issues due to the HAL_GetTick function, I am not very confident to use it...